prusti-encoder/src/encoders/ty/use_impure.rs

```diff
@@ -0,0 +1,380 @@
+    type PrimitiveData = ();
+    type ImmRefData = TyUseImpureImmRef<'vir>;
+    type MutRefData = TyUseImpureMutRef<'vir>;
+    type FieldData = TyUseImpureField<'vir>;
+    type StructData = TyUseImpureStructData<'vir>;
+    type VariantData = ();
+    type EnumData = TyUseImpureEnumData<'vir>;
+}
+
+pub type TyUseImpure<'vir> = Ty<'vir, UseImpureTyDatas>;
+// pub type TyUseImpureData<'vir> = TyData<'vir, UseImpureTyDatas>;
+
+pub type TyUseImpureStruct<'vir> = StructData<'vir, UseImpureTyDatas>;
+pub type TyUseImpureEnum<'vir> = EnumData<'vir, UseImpureTyDatas>;
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyUseImpureData<'vir> {
+    args: GArgsTy<'vir>,
+    impure: <ImpureTyDatas as TyDatas<'vir>>::TyData,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyUseImpureImmRef<'vir> {
+    caster: FieldCaster<'vir>,
+    args: GArgsTy<'vir>,
+    impure: <ImpureTyDatas as TyDatas<'vir>>::ImmRefData,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyUseImpureMutRef<'vir> {
+    caster: FieldCaster<'vir>,
+    args: GArgsTy<'vir>,
+    impure: <ImpureTyDatas as TyDatas<'vir>>::MutRefData,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyUseImpureStructData<'vir> {
+    args: GArgsTy<'vir>,
+    ref_to_pred: PredicateIdn<'vir, (vir::Ref, vir::ManyTyVal, vir::ManyCSnap)>,
+    impure: <ImpureTyDatas as TyDatas<'vir>>::StructData,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyUseImpureField<'vir> {
+    caster: FieldCaster<'vir>,
+    args: GArgsTy<'vir>,
+    impure: <ImpureTyDatas as TyDatas<'vir>>::FieldData,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyUseImpureEnumData<'vir> {
+    args: GArgsTy<'vir>,
+    impure: <ImpureTyDatas as TyDatas<'vir>>::EnumData,
+}
+
+/// Encodes a type into the predicate representation. Takes an arbitrary Rust
+/// `Ty` and provides a wrapper around the results of the `TyImpureEnc` encoder.
+/// This wrapper handles all the generic casts required (e.g. when fold/unfolding).
+pub type TyUseImpureEnc = TyUseEnc<Impure>;
+
+impl TaskEncoder for TyUseImpureEnc {
+    task_encoder::encoder_cache!(TyUseImpureEnc);
+
+    type TaskDescription<'vir> = super::RustTyDecomposition<'vir>;
+
+    type OutputFullDependency<'vir> = TyUseImpure<'vir>;
+
+    fn task_to_key<'vir>(task: &Self::TaskDescription<'vir>) -> Self::TaskKey<'vir> {
+        *task
+    }
+
+    fn do_encode_full<'vir>(
+        task_key: &Self::TaskKey<'vir>,
+        deps: &mut task_encoder::TaskEncoderDependencies<'vir, Self>,
+    ) -> EncodeFullResult<'vir, Self> {
+        deps.emit_output_ref(*task_key, ())?;
+
+        let ty_impure = deps.require_dep::<TyImpureEnc>(task_key.ty)?;
+        let mut walker = TyUseImpureWalker::new(deps, task_key.args);
+        let ty_use_impure = walker.encode_ty(task_key.ty.zip(ty_impure));
+        Ok(((), ty_use_impure.alloc()))
+    }
+
+    fn emit_outputs<'vir>(program: &mut task_encoder::Program<'vir>) {
+        TyImpureEnc::emit_outputs(program)
+    }
+}
+
+struct TyUseImpureWalker<'a, 'vir> {
```

Comment:
I might have missed something: why is this a walker? It doesn't seem to recurse through the structure of a type?

Reply:
It walks the `TyData` structure - it just doesn't recurse into the types of fields. I'm happy to rename it though